port/raspberrypi/rp2xxx/src/hal/i2c_slave.zig

```diff
@@ -0,0 +1,315 @@
+//!
+//! This file implements an I2C slave driver for the RP2 chip family.
+//!
+//! Useful links:
+//!     (Unofficial) I2C Reference: https://www.i2c-bus.org/
+//!
+//! To use this driver, you must open the I2C slave connection
+//! using `I2C_Slave.open()` with the following parameters:
+//!
+//! * `addr` - The I2C slave address to listen to
+//! * `transfer_buffer` - A buffer to use to transfer data
+//! * `rx_callback` - The callback to use for handling data received from i2c
+//! * `tx_callback` - The callback to use for providing data sent to i2c
+//! * `param` - An optional parameter to pass to the callback
+//!
+//! The transfer buffer SHOULD be sized to hold the maximum amount of data
+//! that will be transferred in a single transaction, but it is not required
+//! to be.
+//!
+//! Once the connection is open, and the appropriate ISR is set up,
+//! the driver will listen for I2C transactions.
+//!
+//! ## Master Write / Slave Read
+//!
+//! When we get data from the master we add it to the transfer buffer.  If the
+//! buffer becomes full and we have more data to add, we call the rx_callback
+//! with the `last` parameter set to `false`.
+//!
+//! When we have processed all the data from the master, and the master issues
+//! a STOP or RESTART condition, we call the rx_callback with the `last`
+//! parameter set to `true`.
+//!
+//! In either case, the `first` parameter will be set `true` for the initial
+//! callback for a transaction, and the `gen_call` parameter will be set to
+//! `true` for a general call and `false` if the transaction was addressed to us.
+//!
+//! ## Master Write / Slave Read
+//!
+//! If the master requests data from us we call the tx_callback with the `first`
+//! parameter set to `true`.  The user should add data to `data` and return the
+//! number of bytes added.
+//!
+//! If the master request more data, we call the tx_callback with the `first`
+//! parameter set to `false`.  The user can return 0 if no more data is available.
+//!
+
+const std = @import("std");
+const microzig = @import("microzig");
+const irq = @import("irq.zig");
+
+const mdf = microzig.drivers;
+const peripherals = microzig.chip.peripherals;
+const I2C0 = peripherals.I2C0;
+const I2C1 = peripherals.I2C1;
+const I2cRegs = microzig.chip.types.peripherals.I2C0;
+
+const fifo_length = 16;
+
+const gpio = @import("gpio.zig");
+
+pub const RXCallback = *const fn (data: []const u8, first: bool, last: bool, gen_call: bool, param: ?*anyopaque) void;
```

Comment:
Shouldn't these be named from the perspective of the device we are using? e.g. this is the master RX, but the device is TXing

Reply:
We should probably match the naming used by rpi

https://github.com/raspberrypi/pico-examples/blob/master/i2c/slave_mem_i2c/slave_mem_i2c.c#L40-L51

So RXCallback is called after the equivalent of `I2C_SLAVE_REQUEST`, maybe we call it request callback?

Reply:
Unless I'm looking at this backward, they are.  The TXCallback is called when we, the slave, needs to transmit data, and the RXCallback is called when we receive data from the master.

Reply:
But then how is data const here?

Reply:
RXCallback is called after an equivalent of I2C_SLAVE_RECEIVE.  The master has written some data and we are receiving it.  That's why `data` is const. It is that data we were send and we should not change that.

TXCallback is called after an equivalent of I2C_SLAVE_REQUEST, where the master is requesting data and we need to transmit it.  Here `data` is not const because we do need to put data in the slice.

Reply:
OK :) I had it swapped. The way it is makes sense.